File: Smodr/src/main/java/com/kevintcoughlin/smodr/viewholders/EpisodeView.java
package com.kevintcoughlin.smodr.viewholders;

import android.graphics.Color;
import android.icu.text.SimpleDateFormat;
import android.os.Build;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import com.cascadiacollections.jamoka.adapter.BinderRecyclerAdapter;
import com.kevintcoughlin.smodr.R;
import com.kevintcoughlin.smodr.databinding.ItemListEpisodeLayoutBinding;
import com.kevintcoughlin.smodr.models.Item;
import com.kevintcoughlin.smodr.utils.StringResourceUtilities;

import java.text.ParseException;
import java.util.Date;
import java.util.Locale;

public class EpisodeView implements BinderRecyclerAdapter.Binder<Item, EpisodeViewHolder> {
    // @todo: Date format locale aware
    @RequiresApi(api = Build.VERSION_CODES.N)
    private static final SimpleDateFormat format = new SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss Z", Locale.US);
    // @todo: Date format locale aware
    @RequiresApi(api = Build.VERSION_CODES.N)
    private static final SimpleDateFormat format2 = new SimpleDateFormat("dd MMM", Locale.US);
    // @todo: theme
    private static final int COLOR_BLACK = Color.rgb(0,0,0);
    private static final int COLOR_GRAY = Color.rgb(222,222,222);

    @RequiresApi(api = Build.VERSION_CODES.N)
    private static String formatDate(String dateTimeString) {
        // @todo: optimize
        Date date = null;
        String dateString = "";
        try {
            date = format.parse(dateTimeString);
        } catch (ParseException e) {
            e.printStackTrace();
        }

        if (date != null) {
            dateString = format2.format(date);
        }

        return dateString;
    }

    public EpisodeView() {

    }

    @Override
    public void bind(@NonNull final Item model, @NonNull final EpisodeViewHolder viewHolder) {
        viewHolder.mTitle.setText(model.getTitle());
        viewHolder.mDescription.setText(Html.fromHtml(model.getSummary()));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            viewHolder.mMetadata.setText(StringResourceUtilities.getString(viewHolder.mMetadata.getContext(), R.string.metadata, formatDate(model.getPubDate()), model.getDuration()));
        }

        if (model.getCompleted()) {
            // @todo extension method?
            viewHolder.mTitle.setTextColor(COLOR_GRAY);
            viewHolder.mDescription.setTextColor(COLOR_GRAY);
            viewHolder.mMetadata.setTextColor(COLOR_GRAY);
        } else {
            viewHolder.mTitle.setTextColor(COLOR_BLACK);
            viewHolder.mDescription.setTextColor(COLOR_BLACK);
            viewHolder.mMetadata.setTextColor(COLOR_BLACK);
        }
    }

    @Override
    public EpisodeViewHolder createViewHolder(@NonNull final ViewGroup parent) {
        return new EpisodeViewHolder(ItemListEpisodeLayoutBinding.inflate(LayoutInflater.from(parent.getContext()),
                parent, false));
    }
}

File: Smodr/src/main/java/com/kevintcoughlin/smodr/viewholders/EpisodeView.kt
package com.kevintcoughlin.smodr.viewholders

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.text.HtmlCompat
import com.cascadiacollections.jamoka.adapter.BinderRecyclerAdapter
import com.kevintcoughlin.smodr.R
import com.kevintcoughlin.smodr.databinding.ItemListEpisodeLayoutBinding
import com.kevintcoughlin.smodr.models.Item
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * Binds an Episode (Item) to an EpisodeViewHolder for display in a RecyclerView.
 *
 * This binder uses ViewBinding and Material Design guidelines. It formats episode metadata
 * and summary using HTML compatibility and localized date formatting. The adapter is intended
 * for use with BinderRecyclerAdapter and supports DiffUtil for efficient list updates.
 *
 * @see BinderRecyclerAdapter
 * @see EpisodeViewHolder
 * @see Item
 */
class EpisodeView : BinderRecyclerAdapter.ViewHolderBinder<Item, EpisodeViewHolder> {

    /**
     * Binds the episode data to the view holder's views.
     *
     * @param model The episode item to bind.
     * @param viewHolder The view holder containing the views.
     * @param position The position in the adapter.
     */
    override fun bind(model: Item, viewHolder: EpisodeViewHolder, position: Int) =
        viewHolder.binding.render(model)

    /**
     * Creates a new EpisodeViewHolder using ViewBinding.
     *
     * @param parent The parent ViewGroup.
     * @param viewType The view type (unused).
     * @return A new EpisodeViewHolder instance.
     */
    override fun createViewHolder(parent: ViewGroup, viewType: Int): EpisodeViewHolder {
        val binding = ItemListEpisodeLayoutBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return EpisodeViewHolder(binding)
    }

    private fun ItemListEpisodeLayoutBinding.render(model: Item) {
        title.text = model.title
        description.text = HtmlCompat.fromHtml(model.summary.orEmpty(), HtmlCompat.FROM_HTML_MODE_LEGACY)
        metadata.text = root.resources.getString(
            R.string.metadata,
            formatDate(model.pubDate),
            model.duration.orEmpty(),
        )
    }

    companion object {
        private val dateFormatInput = DateTimeFormatter.RFC_1123_DATE_TIME

        private fun dateFormatOutput(): DateTimeFormatter =
            DateTimeFormatter.ofPattern("dd MMM", Locale.getDefault())

        /**
         * Formats a date string to a localized display format (e.g., "15 Nov").
         * Returns an empty string if parsing fails or input is null/empty.
         *
         * @param dateTimeString The RFC822 date string to format.
         * @return A localized date string or empty string if invalid.
         */
        fun formatDate(dateTimeString: String?): String {
            if (dateTimeString.isNullOrBlank()) return ""
            return runCatching {
                ZonedDateTime.parse(dateTimeString, dateFormatInput).format(dateFormatOutput())
            }.getOrDefault("")
        }
    }
}
